### ISSUE TITLE
Support HTTP DELETE endpoints

### ISSUE THREAD

**Issue by NGA-TRAN** [influxdata/influxdb_iox#2667]
# Rationale
As a part of supporting Delete #2396, we need to accept delete http endpoint like this (thanks @jacobmarble for providing it)

```
POST /api/v2/delete?bucket=mybucket&org=myorg HTTP/1.1
Host: localhost:8086
User-Agent: influx/2.1.0 (darwin) Sha/e6cad1b4aa Date/2021-07-29T17:34:53Z
Transfer-Encoding: chunked
Accept: application/json
Authorization: Token VD5DR9TxyVIa4BoSzNK5I4hTc0b6LGeYjQed4fqW9R3C3FAcvZbeUGkbnxYJGIccCU7LqRYdI9OgdOZlmyymOw==
Content-Type: application/json
Accept-Encoding: gzip

{"predicate":"host=\"Orient.local\"","start":"1970-01-01T00:00:00Z","stop":"2070-01-01T00:00:00Z"}
```
# Design
@jacobmarble has proposed a  write path approach for `deletes` as we did for `line protocols`. Many thanks to @alamb who has pointed us to the implementation of writing line protocol, here is what we are going to do:

Inside IOx's server [route_request](https://github.com/influxdata/influxdb_iox/blob/daa2ec2f4cf3ffe8a97a2e01e5d52b5360870840/src/influxdb_ioxd/http.rs#L364) function, we will add a line to process delete endpoints:

    `(Method::POST, "/api/v2/delete") => delete(req, server.as_ref()).await,`

The `delete` function will look like this
    ```
        async fn delete<M>(
            req: Request<Body>,
            server: &Server<M>,
        ) -> Result<Response<Body>, ApplicationError>
        where
            M: ConnectionManager + Send + Sync + Debug + 'static,
        { .....}
    ```

The request bodies of a delete are JSON object:
   `{"predicate":"host=\"Orient.local\"","start":"1970-01-01T00:00:00Z","stop":"2070-01-01T00:00:00Z"}`

@jacobmarble will provide more examples of request bodies but @e-dard has provided a few input in #2567 which means, in summary, the delete bodies include:

 1. start (required): a timestamp start of the delete range (inclusive)
 2. stop (required): a timestamp end of the delete range (exclusive)
 3. predicate (optional):
       . If empty: we have to delete data between the specified time range for all tables of the DB 
       . includes:
              . zero or one or many table to delete. This is specified by `_measurement` field. If not specified, we have to apply the delete on all tables
              . zero or a conjunctive expression of binary (col eq/not_eq literal)

### A discovery from @jacobmarble 
IDPE has only one delete endpoint, /api/v2/delete. (Write has two: /write for 1.x compatibility and /api/v2/write, and that confuses me often).
How do we support deletes for 1.x compatibility? The [InfluxQL DELETE statement](https://docs.influxdata.com/influxdb/v1.8/query_language/spec/#delete). So there are two paths that support deletes.
The good news is that influxqld implements DELETE statements… :drum_with_drumsticks: … by calling /api/v2/delete! So there is just one delete API to mirror! (edited) 

# Examples:

## Example 1:

        POST /api/v2/delete?bucket=mybucket&org=myorg HTTP/1.1
        .....       
        {"predicate":"host=\"Orient.local\"","start":"1970-01-01T00:00:00Z","stop":"2070-01-01T00:00:00Z"}

is equivalent to 
 
       for every table, T,  in database myorg_mybucket, run
                    DELETE FROM T WHERE host='Orient.local' AND 1970-01-01T00:00:00Z <= time < 2070-01-01T00:00:00Z

## Example 2:

        POST /api/v2/delete?bucket=mybucket&org=myorg HTTP/1.1
        .....
        {"predicate":"_measurement=mytable AND host=\"Orient.local\"","start":"1970-01-01T00:00:00Z","stop":"2070-01-01T00:00:00Z"}

is equivalent to 

      Run the following SQL in database myorg_mybucket
            DELETE FROM mutable WHERE host='Orient.local' AND 1970-01-01T00:00:00Z <= time < 2070-01-01T00:00:00Z

**Comment by alamb** [comment]
It would be cool to get a more formal description of what could be in the delete predicate field.

**Comment by e-dard** [comment]
@alamb https://docs.influxdata.com/influxdb/cloud/reference/syntax/delete-predicate/

**Comment by jacobmarble** [comment]
Thanks for the link @e-dard , very helpful. In particular:

> The delete predicate syntax has the following limitations.
>
> -  Delete predicates do not support regular expressions.
> -  Delete predicates do not support the OR logical operator.
> -  Delete predicates only support equality (=), not inequality (!=).
> -  Delete predicates can use any column or tag except _time or _value.

**Comment by jacobmarble** [comment]
@NGA-TRAN I don't think predicates get more complicated than your Examples 1 and 2, so I don't have any more request body examples to share.

**Comment by NGA-TRAN** [comment]
@jacobmarble : 
- All the limitations were mirrored in IOx and should work the same in IOx today.
- It seems the delete document says nothing about the case of deleting more than 1 specific table.  We either delete from 1 table (e.g. "predicate":"_measurement=cpu") or all tables (e.g. "predicate":""). There are no other cases.

**Comment by alamb** [comment]
The cargo audit failure https://app.circleci.com/pipelines/github/influxdata/influxdb_iox/10880/workflows/27161e28-3b34-459d-adb1-b42e8d2e5e1b/jobs/70289

looks like the same as @mkmik  fixed in conductor. I will port his changes over to fix

**Comment by mkmik** [comment]
> The cargo audit failure https://app.circleci.com/pipelines/github/influxdata/influxdb_iox/10880/workflows/27161e28-3b34-459d-adb1-b42e8d2e5e1b/jobs/70289
> 
> looks like the same as @mkmik fixed in conductor. I will port his changes over to fix

I fixed it in iox:
https://github.com/influxdata/influxdb_iox/pull/2685

**Comment by NGA-TRAN** [comment]
Thanks @alamb . I was about to figure out what was gong on

**Comment by NGA-TRAN** [comment]
> I fixed it in iox: #2685

So I only need to merge main to branch
